python/tvm/relax/training/optimizer.py

```diff
@@ -0,0 +1,582 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Provide abstraction for defining optimizers and a set of common optimizers."""
+
+from decimal import Decimal
+from typing import List, Union
+
+import numpy as np  # type: ignore
+
+import tvm
+from tvm import relax as rx
+from tvm.relax.struct_info import TensorStructInfo
+from tvm.runtime.container import tuple_object
+from tvm.relax.op import add, subtract, multiply, divide, sqrt
+from tvm.relax import Var, Function
+
+
+class Optimizer:
+    """Relax training optimizer. This class could generate relax Functions for optimizing specified
+    parameters, and store the states used in the optimization process, such as momentum.
+
+    Parameters
+    ----------
+    params : Union[Var, List[Var]]
+        The parameter or the list of parameters to optimize.
+
+        If params is None, it indicates params will be added later using add_params.
+
+        Parameters should all be Vars of floating point Tensors, including float32, float64,
+        float16, bfloat16, etc. Currently, all parameters should have the same dtype, and that dtype
+        will be used as the dtype of the optimizer states.
+    """
```

Comment:
Let’s add some doc strings for the param list, state and dtype. Specifically about `state`: what could be the contents of the state, etc. (I’m new to optimizers and this is the first question popped out of my head.)
```suggestion
    Attributes
    ----------
    ...
    """
```

Reply:
We don't want users use underscored attributes directly. They would be used only in subclasses of Optimizer. Should we still add documents for underscored attributes?

Reply:
I see. In this case no need to document them in attribute section, but definitely explain something about these three somewhere in the doc.

Reply:
I see the document for state in the property below. So here you can refer to the document there.

Reply:
Sure, I would add that